### PR TITLE
[RFR] RAT builder rework

### PR DESCRIPTION
**Pull request by DylanBell**
Improves rat-builder functionality and resolves a number of bugs

- Reworks the way that we create and persist choices and questions
- Modifies the way that we set assignment `release_at` and `due_at`
- Reworks the way assessments are synced
- Various client-side validation improvements
- And more!

### CODE REVIEW

File: src/thinkspace/api/thinkspace-casespace/app/models/thinkspace/casespace/assignment.rb [hunk 1]
```diff
@@ -234,7 +234,7 @@ def delete_ownerable_data(ownerables)
         end
       end
 
-      def sync_rat
+      def sync_rat?
         return false unless settings.present? && settings.dig('rat', 'sync').present?
         return settings['rat']['sync']
```
Comment (jamesfid) [inline src/thinkspace/api/thinkspace-casespace/app/models/thinkspace/casespace/assignment.rb]: This should be a `dig`, especially if using it in the above line.

File: src/thinkspace/api/thinkspace-readiness-assurance/app/models/thinkspace/readiness_assurance/assessment.rb [hunk 2]
```diff
@@ -1,7 +1,7 @@
 module Thinkspace
   module ReadinessAssurance
     class Assessment < ActiveRecord::Base
-      after_save :sync_assessments
+      #after_save :sync_assessments
```
Comment (jamesfid) [inline src/thinkspace/api/thinkspace-readiness-assurance/app/models/thinkspace/readiness_assurance/assessment.rb]: This line should be removed if we're opting for another method.

File: src/thinkspace/api/thinkspace-readiness-assurance/app/models/thinkspace/readiness_assurance/assessment.rb [hunk 3]
```diff
@@ -29,14 +29,109 @@ def sync_assessments
...
           assessments = Thinkspace::ReadinessAssurance::Assessment.where(authable: assignment.thinkspace_casespace_phases).without(self)
           assessments.each do |assessment|
             assessment.update_columns(questions: self.questions, answers: self.answers, settings: self.settings.deep_merge(assessment.settings))
           end
         end
       end
 
+      def sync(options)
+        phase      = self.authable
+        assignment = Thinkspace::Casespace::Assignment.find(phase.assignment_id)
```
Comment (jamesfid) [inline src/thinkspace/api/thinkspace-readiness-assurance/app/models/thinkspace/readiness_assurance/assessment.rb]: phase.thinkspace_casespace_assignment

File: src/thinkspace/api/thinkspace-readiness-assurance/app/models/thinkspace/readiness_assurance/assessment.rb [hunk 4]
```diff
@@ -29,14 +29,109 @@ def sync_assessments
...
+
+        scoring_keys  = ['correct', 'attempted', 'no_answer', 'incorrect_attempt']
+        question_keys = ['type', 'random', 'ifat', 'justification']
+
+        if assignment.sync_rat?
```
Comment (jamesfid) [inline src/thinkspace/api/thinkspace-readiness-assurance/app/models/thinkspace/readiness_assurance/assessment.rb]: This is some pretty complicated logic.  Couple of notes:
  * It'd be nice to pull this out into a class that can not pollute the model as much.  There's a lot of nested if/if statements that would be nice to have as separate method (which lends well towards the class abstraction).  For example, you could have a method called `sync_settings` which handles just the settings, etc.
  * I'd use `dig` pretty much everywhere.  It just feels a lot safer than running into `undefined [] for nil` type of error.  I know most of these are scoped out by the if wrappers.

File: src/thinkspace/client/thinkspace-assets/styles/otbl-css/_components.forms.scss [hunk 5]
```diff
@@ -10,13 +10,15 @@ fieldset {
 
 .forms__section {
 	>h2 {
-		font-size: 1.15rem;
+		// font-size: 1.15rem;
```
Comment (jamesfid) [inline src/thinkspace/client/thinkspace-assets/styles/otbl-css/_components.forms.scss:13]: Remove if not needed.

File: src/thinkspace/client/thinkspace-assets/styles/otbl-css/_helpers.spacing.scss [hunk 6]
```diff
@@ -95,7 +95,7 @@
  */
 
 .spacing__margin-top--0 {
-  margin-top: 0;
+  margin-top: 0 !important;
```
Comment (jamesfid) [inline src/thinkspace/client/thinkspace-assets/styles/otbl-css/_helpers.spacing.scss:98]: Did @jcamou add this?  Not a fan of `!important` if it's avoidable.

File: src/thinkspace/client/thinkspace-assets/styles/otbl-css/_objects.checkbox.scss [hunk 7]
```diff
@@ -2,72 +2,173 @@
 //    #Checkbox
 //\*------------------------------------*/
 
+// .checkbox__item {
```
Comment (jamesfid) [inline src/thinkspace/client/thinkspace-assets/styles/otbl-css/_objects.checkbox.scss:5]: Removed commented code if not needed.

File: src/thinkspace/client/thinkspace-builder-rat/addon/components/content/question/item/show.coffee [hunk 8]
```diff
@@ -13,15 +13,16 @@ export default base.extend
...
-
-    delete:    -> @get('manager').delete_question_item(@get('type'), @get('model.model'))
-
-    reorder_up:     -> @get('manager').reorder_item(@get('type'), @get('model.model'), -1)
-    reorder_down:   -> @get('manager').reorder_item(@get('type'), @get('model.model'), 1)
-    reorder_top:    -> @get('manager').reorder_item(@get('type'), @get('model.model'), 'top')
-    reorder_bottom: -> @get('manager').reorder_item(@get('type'), @get('model.model'), 'bottom')
\ No newline at end of file
+    duplicate:      -> @get('step').duplicate_question_item(@get('type'), @get('model.model'))
+    delete:         ->
```
Comment (jamesfid) [inline src/thinkspace/client/thinkspace-builder-rat/addon/components/content/question/item/show.coffee]: Make this a one liner like the rest.

File: src/thinkspace/client/thinkspace-builder-rat/addon/items/question.coffee [hunk 9]
```diff
@@ -84,29 +93,22 @@ export default ember.Object.extend array_helpers,
...
     vpresence = totem_changeset.vpresence(true)
 
     choices   = @duplicate_array(@get('model.choices'))
     changeset = totem_changeset.create(model,
       question: [vpresence],
-      answer:   [vpresence],
       choices:  [vpresence]
     )
 
+    answer_cs = totem_changeset.create(@,
```
Comment (jamesfid) [inline src/thinkspace/client/thinkspace-builder-rat/addon/items/question.coffee]: If you don't have the leading ( you won't need the closing ).

File: src/thinkspace/client/thinkspace-builder-rat/addon/services/manager.coffee [hunk 10]
```diff
@@ -50,83 +161,126 @@ export default ember.Service.extend array_helpers, uuid,
...
-    add_at = index + 1
-    return if add_at < 0
-    new_item = ember.merge({}, item)
-    new_item.id = @get_next_id(type)
-    items.insertAt add_at, new_item
+    new ember.RSVP.Promise (resolve, reject) =>
+      items = @get_items(type)
+      item  = items.findBy('id', item.id)
+      index = items.indexOf(item)
+      console.log('[DUP] index is ', index)
```
Comment (jamesfid) [inline src/thinkspace/client/thinkspace-builder-rat/addon/services/manager.coffee]: Remove.

File: src/thinkspace/client/thinkspace-builder-rat/addon/services/manager.coffee [hunk 11]
```diff
@@ -50,83 +161,126 @@ export default ember.Service.extend array_helpers, uuid,
...
+      resolve() unless index > -1
+      add_at = index + 1
+      console.log('[DUP] add_at is ', add_at)
```
Comment (jamesfid) [inline src/thinkspace/client/thinkspace-builder-rat/addon/services/manager.coffee]: Remove.

File: src/thinkspace/client/thinkspace-builder-rat/addon/steps/content.coffee [hunk 12]
```diff
@@ -74,15 +87,10 @@ export default step.extend
...
-  irat_question_items: ember.computed 'irat_assessment.questions_with_answers.@each', ->
-    items = @get('irat_assessment.questions_with_answers')
-    if ember.isPresent(items)
-      @create_question_item(@get('irat_type'), item) for item in items
-
-  trat_question_items: ember.computed 'trat_assessment.questions_with_answers.@each', ->
-    items = @get('trat_assessment.questions_with_answers')
-    if ember.isPresent(items)
-      @create_question_item(@get('trat_type'), item) for item in items
+  # trat_question_items: ember.computed 'trat_assessment.questions_with_answers.@each', ->
```
Comment (jamesfid) [inline src/thinkspace/client/thinkspace-builder-rat/addon/steps/content.coffee:88]: Remove commented code if not needed.

File: src/thinkspace/client/thinkspace-builder-rat/addon/steps/content.coffee [hunk 13]
```diff
@@ -107,8 +115,58 @@ export default step.extend
...
+    manager = @get('manager')
+    @set_loading('irat')
+    manager.duplicate_question_item(type, item_obj).then =>
+      @create_question_items()
+      @reset_loading('irat')
+
+  reorder_up: (type, item_obj) ->
+    manager = @get('manager')
+    @set_loading('irat')
+    manager.reorder_item(type, item_obj, -1).then =>
```
Comment (jamesfid) [inline src/thinkspace/client/thinkspace-builder-rat/addon/steps/content.coffee]: Feels like all of these `manager.reorder_item` calls should be abstracted out.  Lots of repeat code in the `then` handler. Something like:

```
  @manager_reorder(type, obj, offset)
```

File: src/thinkspace/client/thinkspace-builder-rat/addon/steps/details.coffee [hunk 14]
```diff
@@ -17,77 +17,115 @@ export default step.extend
...
-      tc.query_action(ta.to_p('assignment'), params, options).then (assignment) =>
-        resolve assignment
-
-  query_team_sets: ->
-    new ember.RSVP.Promise (resolve, reject) =>
-      @get('model').get(ta.to_p('space')).then (space) =>
-        space.get_default_team_set().then (team_set) =>
-          @set 'team_sets', ember.makeArray(team_set)
-          resolve()
+  # load_assignment_data: ->
```
Comment (jamesfid) [inline src/thinkspace/client/thinkspace-builder-rat/addon/steps/details.coffee]: Removed if not needed.

File: src/thinkspace/client/thinkspace-builder-rat/addon/steps/details.coffee [hunk 15]
```diff
@@ -17,77 +17,115 @@ export default step.extend
...
+  #       resolve assignment
 
-  select_team_set: (team_set) -> 
-    # TODO: This needs to be refactored to get the tRAT assessment, then get the authable for it.
-    @set 'selected_team_set', team_set
-    @get('model').get(ta.to_p('phases')).then (phases) =>
-      phase = phases.get('lastObject')
-      phase.set 'team_set_id', team_set.get('id')
-      phase.save()
+  # query_team_sets: ->
```
Comment (jamesfid) [inline src/thinkspace/client/thinkspace-builder-rat/addon/steps/details.coffee]: Remove if not needed.

File: src/thinkspace/client/thinkspace-builder-rat/addon/steps/details.coffee [hunk 16]
```diff
@@ -17,77 +17,115 @@ export default step.extend
...
-        team_sets   = @get('team_sets')
-        phase       = phases.get('firstObject')
-        if ember.isPresent(phase.get('team_set_id')) 
-          team_set = team_sets.findBy 'id', phase.get('team_set_id').toString()
-          @set 'selected_team_set', team_set
-        else 
-          team_set = team_sets.get('firstObject')
-          @select_team_set team_set
-        resolve()
+  # select_team_set: (team_set) ->
```
Comment (jamesfid) [inline src/thinkspace/client/thinkspace-builder-rat/addon/steps/details.coffee]: Remove if not ndeeded.

File: src/thinkspace/client/thinkspace-builder-rat/addon/steps/details.coffee [hunk 17]
```diff
@@ -17,77 +17,115 @@ export default step.extend
...
-        resolve()
+  # select_team_set: (team_set) -> 
+  #   # TODO: This needs to be refactored to get the tRAT assessment, then get the authable for it.
+  #   @set 'selected_team_set', team_set
+  #   @get('model').get(ta.to_p('phases')).then (phases) =>
+  #     phase = phases.get('lastObject')
+  #     phase.set 'team_set_id', team_set.get('id')
+  #     phase.save()
+
+  # initialize_team_set: ->
```
Comment (jamesfid) [inline src/thinkspace/client/thinkspace-builder-rat/addon/steps/details.coffee]: Remove if not needed.

File: src/thinkspace/client/thinkspace-builder-rat/addon/steps/settings.coffee [hunk 18]
```diff
@@ -18,73 +20,152 @@ export default step.extend
...
+    irat_changeset = totem_changeset.create(irat_assessment)
+    trat_changeset = totem_changeset.create(trat_assessment)
 
-      query =
-        id: model.get('id')
-        componentable_type: ns.to_p('ra:assessment')
-      options =
-        action: 'phase_componentables'
-        model: ns.to_p('ra:assessment')
+    scoring_changeset = totem_changeset.create(irat_assessment.get('settings.scoring'),
```
Comment (jamesfid) [inline src/thinkspace/client/thinkspace-builder-rat/addon/steps/settings.coffee]: Some comment as earlier regarding parens.

File: src/thinkspace/client/thinkspace-builder-rat/addon/steps/settings.coffee [hunk 19]
```diff
@@ -18,73 +20,152 @@ export default step.extend
...
+      correct:           [v_integer, v_presence, v_positive],
+      no_answer:         [v_integer, v_presence],
+      #attempted:         [v_integer, v_presence, v_positive],
```
Comment (jamesfid) [inline src/thinkspace/client/thinkspace-builder-rat/addon/steps/settings.coffee]: Remove commented portions if not used.

File: src/thinkspace/client/thinkspace-builder-rat/addon/steps/settings.coffee [hunk 20]
```diff
@@ -18,73 +20,152 @@ export default step.extend
...
+    )
 
-  init_assessments: ->
-    assessments = @get('assessments')
-    manager     = @get('manager')
+    @init_incorrect_attempt(trat_scoring_cs)
 
-    irat = assessments.findBy 'is_irat', true
-    trat = assessments.findBy 'is_trat', true
+    changeset.set('show_errors', true)
```
Comment (jamesfid) [inline src/thinkspace/client/thinkspace-builder-rat/addon/steps/settings.coffee]: There is already a function for this, something like `show_errors_on()` - look at the changeset file.

File: src/thinkspace/client/thinkspace-builder-rat/addon/steps/settings.coffee [hunk 21]
```diff
@@ -18,73 +20,152 @@ export default step.extend
...
+    changeset  = @get('changeset')
+
+    unlock_at = if irat_phase.get('unlock_at') <= trat_phase.get('unlock_at') then irat_phase.get('unlock_at') else trat_phase.get('unlock_at')
+    due_at    = if irat_phase.get('due_at')    >= trat_phase.get('due_at')    then irat_phase.get('due_at')    else trat_phase.get('due_at')
+
+    changeset.set('release_at', unlock_at)
+    changeset.set('due_at',     due_at)
+
+  toggle_rat_changeset_property: (type, property) ->
+    #changeset = @get('irat_changeset')
```
Comment (jamesfid) [inline src/thinkspace/client/thinkspace-builder-rat/addon/steps/settings.coffee]: Remove.

**Review by jamesfid — CHANGES_REQUESTED** [review]

File: src/thinkspace/api/thinkspace-casespace/app/models/thinkspace/casespace/assignment.rb (hunk 1 above)
Comment (DylanBell) [reply src/thinkspace/api/thinkspace-casespace/app/models/thinkspace/casespace/assignment.rb]: Addressed

File: src/thinkspace/api/thinkspace-readiness-assurance/app/models/thinkspace/readiness_assurance/assessment.rb (hunk 2 above)
Comment (DylanBell) [reply src/thinkspace/api/thinkspace-readiness-assurance/app/models/thinkspace/readiness_assurance/assessment.rb]: Addressed

File: src/thinkspace/api/thinkspace-readiness-assurance/app/models/thinkspace/readiness_assurance/assessment.rb (hunk 3 above)
Comment (DylanBell) [reply src/thinkspace/api/thinkspace-readiness-assurance/app/models/thinkspace/readiness_assurance/assessment.rb]: Addressed

File: src/thinkspace/client/thinkspace-assets/styles/otbl-css/_components.forms.scss (hunk 5 above)
Comment (DylanBell) [reply src/thinkspace/client/thinkspace-assets/styles/otbl-css/_components.forms.scss:13]: Modified by @jcamou

File: src/thinkspace/client/thinkspace-assets/styles/otbl-css/_helpers.spacing.scss (hunk 6 above)
Comment (DylanBell) [reply src/thinkspace/client/thinkspace-assets/styles/otbl-css/_helpers.spacing.scss:98]: Modified by @jcamou

File: src/thinkspace/client/thinkspace-assets/styles/otbl-css/_objects.checkbox.scss (hunk 7 above)
Comment (DylanBell) [reply src/thinkspace/client/thinkspace-assets/styles/otbl-css/_objects.checkbox.scss:5]: Modified by @jcamou

File: src/thinkspace/client/thinkspace-builder-rat/addon/components/content/question/item/show.coffee (hunk 8 above)
Comment (DylanBell) [reply src/thinkspace/client/thinkspace-builder-rat/addon/components/content/question/item/show.coffee]: Addressed

File: src/thinkspace/client/thinkspace-builder-rat/addon/services/manager.coffee (hunk 10 above)
Comment (DylanBell) [reply src/thinkspace/client/thinkspace-builder-rat/addon/services/manager.coffee]: Addressed

File: src/thinkspace/client/thinkspace-builder-rat/addon/services/manager.coffee (hunk 11 above)
Comment (DylanBell) [reply src/thinkspace/client/thinkspace-builder-rat/addon/services/manager.coffee]: Addressed

File: src/thinkspace/client/thinkspace-builder-rat/addon/steps/content.coffee (hunk 12 above)
Comment (DylanBell) [reply src/thinkspace/client/thinkspace-builder-rat/addon/steps/content.coffee:88]: Something that's in the process of being modified in the yet-unpushed `feature/rat-desync` branch, to support re-adding the iRAT/tRAT desync capability

File: src/thinkspace/client/thinkspace-assets/styles/otbl-css/_helpers.spacing.scss (hunk 6 above)
Comment (jcamou) [reply src/thinkspace/client/thinkspace-assets/styles/otbl-css/_helpers.spacing.scss:98]: I've only added !important to the helper classes. Helper class attributes need to take priority

File: src/thinkspace/client/thinkspace-builder-rat/addon/steps/details.coffee (hunk 15 above)
Comment (DylanBell) [reply src/thinkspace/client/thinkspace-builder-rat/addon/steps/details.coffee]: Addressed

File: src/thinkspace/client/thinkspace-builder-rat/addon/steps/details.coffee (hunk 14 above)
Comment (DylanBell) [reply src/thinkspace/client/thinkspace-builder-rat/addon/steps/details.coffee]: Addressed

File: src/thinkspace/client/thinkspace-builder-rat/addon/steps/details.coffee (hunk 16 above)
Comment (DylanBell) [reply src/thinkspace/client/thinkspace-builder-rat/addon/steps/details.coffee]: Addressed

File: src/thinkspace/client/thinkspace-builder-rat/addon/steps/details.coffee (hunk 17 above)
Comment (DylanBell) [reply src/thinkspace/client/thinkspace-builder-rat/addon/steps/details.coffee]: Addressed

File: src/thinkspace/client/thinkspace-builder-rat/addon/steps/settings.coffee (hunk 19 above)
Comment (DylanBell) [reply src/thinkspace/client/thinkspace-builder-rat/addon/steps/settings.coffee]: Addressed

File: src/thinkspace/client/thinkspace-builder-rat/addon/steps/settings.coffee (hunk 21 above)
Comment (DylanBell) [reply src/thinkspace/client/thinkspace-builder-rat/addon/steps/settings.coffee]: Addressed

File: src/thinkspace/client/thinkspace-builder-rat/addon/steps/settings.coffee (hunk 20 above)
Comment (DylanBell) [reply src/thinkspace/client/thinkspace-builder-rat/addon/steps/settings.coffee]: Addressed

File: src/thinkspace/client/thinkspace-builder-rat/addon/steps/settings.coffee (hunk 18 above)
Comment (DylanBell) [reply src/thinkspace/client/thinkspace-builder-rat/addon/steps/settings.coffee]: Addressed

File: src/thinkspace/client/thinkspace-builder-rat/addon/items/question.coffee (hunk 9 above)
Comment (DylanBell) [reply src/thinkspace/client/thinkspace-builder-rat/addon/items/question.coffee]: Addressed

File: src/thinkspace/api/thinkspace-readiness-assurance/app/models/thinkspace/readiness_assurance/assessment.rb (hunk 4 above)
Comment (DylanBell) [reply src/thinkspace/api/thinkspace-readiness-assurance/app/models/thinkspace/readiness_assurance/assessment.rb]: Reworked by creating the `Thinkspace::ReadinessAssurance::Sync::Assessment` class, which is created by the `Thinkspace::ReadinessAssurance::Assessment` `sync` method. Currently the same `sync` method functionality is simply transplanted to the `Sync::Assessment`

File: src/thinkspace/client/thinkspace-builder-rat/addon/steps/content.coffee (hunk 13 above)
Comment (DylanBell) [reply src/thinkspace/client/thinkspace-builder-rat/addon/steps/content.coffee]: Addressed

**Pull request merged**